### ISSUE TITLE
Allow loading CSS after the page loads

### ISSUE THREAD

**Issue by mrclay** [Elgg/Elgg#8405]
Goal is simple but CSS is not, primarily due to the cascade that makes order of the sheets in the doc critical. Whether we load them as modules or allow `elgg_load_css` to work client-side, order will be an issue.

One area of thinking is to allow responses in the [new Ajax API](https://github.com/Elgg/Elgg/issues/8323#issuecomment-107914707) to tell the client what CSS to load (and I guess maybe where to insert it in the doc).

See last year's presentation on [Instagram's web stack](http://m.youtube.com/watch?v=VkTCL6Nqm6Y) (webpack) for complications with the cascade.

**Comment by mrclay** [comment]
Particularly for things like lightbox, a plugin will be tempted to write CSS to override colorbox CSS. If we lazy-load the colorbox CSS and it goes later in the document, it could easily break the customizations. IIRC Instagram encouraged some flavor of discipline in how CSS was written:/

**Comment by hypeJunction** [comment]
I have been experimenting with deferred CSS loading: https://github.com/hypeJunction/Elgg-defer_css. Comments are welcome.

**Comment by mrclay** [comment]
I think ideally we'd use a Require loader, which would allowing setting CSS views as real module dependents. E.g. https://github.com/guybedford/require-css

**Comment by mrclay** [comment]
On the server side I've started work altering the elgg_require_js (and adding elgg_require_css) so that they send back a list of modules with `elgg/Ajax` requests, and the Ajax module would require them client-side. A small blocker to this plan is that the calling client-side code would want to know when those modules/CSS had been loaded.

**Comment by mrclay** [comment]
...and this leaves a small race condition. Let's say you pulled a view:
```js
ajax.view('foo/bar').done(function (html) {
  // if this view used elgg_require_js(), then a module might be loading now
  // but we don't know when or if it succeeds. If that module arrives before this
  // html gets in the DOM, that could be a problem, too.
});
```

**Comment by hypeJunction** [comment]
Can we not resolve a promise until all required modules have been loaded?

**Comment by mrclay** [comment]
Yes, we could. I was thinking about trying to do this like a plugin would (just hooks). Ideally elgg/Ajax hooks could implement some kind of blocking like this.

**Comment by mrclay** [comment]
...not make the hook synchronous, but give the elgg/Ajax response a little note to wait on a promise.

**Comment by hypeJunction** [comment]
On the other hand, if they know what exact module they needs, they can just require it again. I think the main problem is that the requester doesn't really know what other plugins are doing.

**Comment by mrclay** [comment]
Yeah, no way to get around that. Simplest would be to recommend sticking inline `<script>require(['foo', 'css!foo'])</script>` into views. When they hit the DOM, require just handles it. :/

**Comment by mrclay** [comment]
Ultimately comes down to how far we want to go down the path of building apps by shoving HTML into the DOM. Seems OK for trivial little components at least.

**Comment by hypeJunction** [comment]
That is what I am doing and I don't really see a point in investing endless
resources into a "proper" solution.

**Comment by jdalsem** [comment]
Closing this as there will be no solution on the horizon where we have control over order or know if css is already loaded

See #11367
